docs: add blog on using duckdb for RAG (#8387)
ibis-project/ibis: docs/posts/duckdb-for-rag/index.qmd

```

Let's jump into the code!

## RAG over Ibis documentation

We'll use the Ibis documentation as our knowledge base.

### Setup code

First we need to install and import the necessary packages:

```bash
pip install 'ibis-framework[duckdb]' openai python-dotenv
```

```{python}
import os  # <1>
import ibis
import shutil
import subprocess

from openai import OpenAI
from dotenv import load_dotenv
from pathlib import Path  # <1>
```

1. Import the libraries we'll use.

We need to set the `OPENAI_API_KEY` environment variable to our OpenAI API key.
Do this by creating a `.env` file in the same directory as the code and adding
the following line:

```txt
OPENAI_API_KEY=<your-api-key>  # <1>
```

1. Replace with your OpenAI API key.

Then load the environment variables:

```{python}
load_dotenv()  # <1>
```

1. Load environment variables from the `.env` file.

Now we're ready to make OpenAI calls.

Let's also setup Ibis for interactive use:

```{python}
ibis.options.interactive = True  # <1>
```

1. Turn on Ibis interactive mode.

### Getting the data into DuckDB

We'll use a function to clone a given repository:

```{python}
def clone_repo(repo_url):  # <1>

>     subprocess.run(["git", "clone", "--depth", "1", repo_url], check=False)  # <1>

Reviewer: Can't you clone to `tempfile.gettempdir()` instead of having to delete using `rmtree`?

Author: why do you think I would know these things. ima PM

Reviewer: Why clone at all even? Couldn't you refer to your local editable module?

```python
Path(ibis.__file__).parents[1].joinpath("docs").rglob("*.qmd")
```

Author: yes but wanted to make it easy for someone to copy and run w/ their own repo (just change the repo url/name)

Author: I'll change to that though and just add a tip about changing the whatever to your own docs

---

> title: "Using DuckDB (or any database Ibis supports) for RAG"

Author: I want to keep the DuckDB hype, and it's obviously using Ibis (on the Ibis website)

what about just "Using DuckDB for RAG"? I think I had that initially -- then it's pointed out right at the top you can do it w/ (practically) any database

> title: "Using DuckDB (or any database Ibis supports) for RAG"

Author: or "Using DuckDB for RAG in Python"

> title: "Using DuckDB (or any database Ibis supports) for RAG"

Author: then the reader's like DuckDB is SQL only, how can this be Python

> title: "Using DuckDB (or any database Ibis supports) for RAG"

Reviewer: I do think there's value in restating "Ibis" in blog post titles, even if it's only for search indexes.

How about 

Using Ibis + DuckDB for RAG

> title: "Using DuckDB (or any database Ibis supports) for RAG"

Author: "Using DuckDB + Ibis for RAG", final offer

> title: "Using DuckDB (or any database Ibis supports) for RAG"

Reviewer: I guess I'll just live with the fact that you like DuckDB more than Ibis 😭

### Getting the data into DuckDB

We'll use a function to clone a given repository:

```{python}
def clone_repo(repo_url):  # <1>
    subprocess.run(["git", "clone", "--depth", "1", repo_url], check=False)  # <1>
```

1. Define a function to clone a repository.

Let's clone `ibis` and prepare to ingest all the Quarto markdown files into a
DuckDB database called `rag.ddb`:

```{python}
org_url = "https://github.com/ibis-project"  # <1>
repo_name = "ibis"  # <1>

shutil.rmtree(repo_name, ignore_errors=True)  # <2>
clone_repo(f"{org_url}/{repo_name}")  # <3>

if os.path.exists("rag.ddb"):  # <4>
    os.remove("rag.ddb")
if os.path.exists("rag.ddb.wal"):
    os.remove("rag.ddb.wal")  # <4>

rag_con = ibis.connect("duckdb://rag.ddb")  # <5>
rag_con.list_tables()  # <6>
```

1. Setup the repository to clone.
2. Remove the repository if it exists to ensure a fresh clone.
3. Clone the repository.
4. Remove the DuckDB database if it exists to ensure a fresh database.
5. Connect to the DuckDB database.
6. List the tables in the database, showing it is empty.

We'll walk through all the cloned files. For `.qmd` files, read the contents and
insert them into the `ibis` table:

```{python}

> for root, _, files in os.walk(repo_name):  # <1>

Reviewer: I think `Path(repo_name).rglob("*.qmd")` might be a less cluttered approach to recursively reading all `.qmd` files.

    for file in files:  # <2>
        if file.endswith(".qmd"):  # <3>
            with open(os.path.join(root, file)) as f:  # <4>
                contents = f.read()  # <5>

>                 data = {  # <6>
>                     "filepath": [os.path.join(root, file)],
>                     "contents": [contents],
>                 }  # <6>
> 
>                 t = ibis.memtable(data)  # <7>
> 
>                 if repo_name not in rag_con.list_tables():  # <8>
>                     rag_con.create_table(repo_name, t)  # <8>
>                 else:  # <9>
>                     rag_con.insert(repo_name, t)  # <9>

Reviewer: ```suggestion
            data = {  # <6>
                "filepath": [os.path.join(root, file)],
                "contents": [contents],
            }  # <6>

            t = ibis.memtable(data)  # <7>

            if repo_name not in rag_con.list_tables():  # <8>
                rag_con.create_table(repo_name, t)  # <8>
            else:  # <9>
                rag_con.insert(repo_name, t)  # <9>
```

Should be able to dedent all these lines and only keep the file open for as long as necessary, and reduce nesting a bit

for now. Let's augmented our data with the estimated number of tokens and sort
from largest to smallest:

```{python}
@ibis.udf.scalar.python  # <1>
def estimate_tokens(text: str) -> int:  # <1>
    return len(text) // 4  # <1>
```

1. Create an Ibis UDF using the Python function.

```{python}
t = (
    t.mutate(tokens_estimate=estimate_tokens(t.contents))  # <1>
    .order_by(ibis._["tokens_estimate"].desc()) # <2>
    .relocate("filepath", "tokens_estimate") # <3>
)
t
```

1. Augment the table with the estimated number of tokens.
2. Sort the table by the estimated number of tokens in descending order.
3. Relocate the `filepath` and `tokens_estimate` columns to the front of the table.

The longest text is certainly over our embedding model's token limit, and the
second longest might be as well. In practice, we'd want to chunk our text into
smaller pieces. This is its own challenge, but for the purposes of this
demonstration we'll just ignore text that is too long.

Let's define our embedding functions:

```{python}
def _embed(text: str) -> list[float]:  # <1>
    """Text to fixed-length array embedding."""  # <1>

>     try:  # <2>
>         model = "text-embedding-3-small"  # <3>
>         text = text.replace("\n", " ")  # <4>
>         client = OpenAI()  # <5>
>         return (  # <6>
>             client.embeddings.create(input=[text], model=model).data[0].embedding
>         )  # <6>
>     except Exception as e:  # <7>
>         print(e)  # <7>
>         return None  # <7>

Reviewer: ```suggestion
    model = "text-embedding-3-small"  # <3>
    text = text.replace("\n", " ")  # <4>
    client = OpenAI()  # <5>
    try:  # <2>
        return (  # <6>
            client.embeddings.create(input=[text], model=model).data[0].embedding
        )  # <6>
    except Exception as e:  # <7>
        print(e)  # <7>
        return None  # <7>
```

Seems like the intent of catching the exception is to deal with failed API calls. Moving the client setup and call argument setup out of the except block might help clarify that.

@ibis.udf.scalar.python  # <8>
def embed(text: str, tokens_estimate: int) -> list[float]:  # <8>
    """Text to fixed-length array embedding."""  # <8>

>     if 0 < tokens_estimate < 8191:  # <9>

Reviewer: Why the hardcoded 8191? Is that covered elsewhere in the post?

Author: it's noted above, that's the max token length allowed by the embedding models

Reviewer: Suspiciously close to 8192, wonder what that extra 1 is for 🤷🏻

Over time, these context lengths have been increasing -- 16k, then 32k, and now
128k for GPT-4-turbo models. Other LLMs have even higher context lengths, with
Google's Gemini family reporting 1M and 10M token context lengths. However, even
these large context lengths do not solve some key problems. For one, LLMs may
ignore or even be confused by context. While you can feed in a full textbook on
machine learning, if you're just asking for information on k-means clustering
you should probably only include that chapter. Additionally, you are charged
money by the number of input tokens (and response tokens) and the time it takes
to inference is proportional to that number. Regardless of the context length of
the model, for the foreseeable future you want to be as efficient as possible.

Thus, you need to have some method of retrieving and augmentign your input text
with the most relevant information for the task at hand.

## Setup

Another reason I dislike RAG is, like traditional ML (and even software
engineering), is that the best approach is often "**it depends**". What are you
trying to accomplish? What are your constraints on resources?

### Scenario

> For this blog, we'll use a scenario: we are trying to build a state-of-the-art

Reviewer 2: You already have "scenario" on the title
```suggestion
For this blog, we are going to try to build a state-of-the-art
```

exploratory data analysis (EDA) chatbot. I want to be able to ask it questions
about my data, have it generate accurate SQL, and run that SQL on my behalf.

We'll be using toy data (our favorite penguins dataset!) and LLM-generated
English and SQL queries, but the same principles should apply for real-world use
cases.

### Setup code

> First, we need to install the necessary packages:

Reviewer 2: ```suggestion
First, we need to install and import the necessary packages:
```

```bash
pip install 'ibis-framework[duckdb,examples]' openai marvin python-dotenv
```

> Then import them:

Reviewer 2: ```suggestion
```

```{python}
import ibis
import marvin

from openai import OpenAI

from dotenv import load_dotenv
```

> You need to set the `OPENAI_API_KEY` environment variable to your OpenAI API

Reviewer 2: In the install section, we are talking in the "We" person and here you say "You" need, you want to pick one and be consistent. Or remove the person entirely.

```

and get our penguins:

```{python}
penguins = ibis.examples.penguins.fetch()
penguins.limit(3)
```

## Demonstration

We'll start with a slightly modified text-to-SQL function [stolen from our
previous "LMs for data" blog](../lms-for-data/index.qmd):

```{python}
@marvin.fn
def _text_to_sql(
    text: str,
    table_name: str,
    table_schema: str,
    table_preview: str,
) -> str:
    """Writes a SQL SELECT statement for the `text` given the provided
    `table_name`, `table_schema`, and `table_preview`.

    Spell out all column names, never use SELECT *. Include ALL columns unless told otherwise.
    Use newlines and indentation for readability. Use LIMIT, not TOP.
    Select FROM the `table_name` only (penguins), you have no access to other tables.
    """

def text_to_sql(
    text: str,
    table_name: str,
    table_schema: str,
    table_preview: str,
) -> str:
    sql = _text_to_sql(text, table_name, table_schema, table_preview)
    return sql.strip().strip(";")
```

This will take the user's English (or really, any language the LLM sufficiently
knows) query and turn it into SQL.

:::{.callout-info title="Isn't this already RAG?"}

> We have retrieved information (a table name, schema, and preview) augmented our

Reviewer 2: This reads weird, it's missing some connector, not sure what are you trying to say though.
"We have retrieved information augmented our..."

model's input with that information, and generated a response.
:::

Our hypothetical service takes text as input, generates SQL, and runs it against
the database on behalf of the user. We need some metric to evaluate the user's
feedback on the generated SQL. This could be an explicit input from the user
(rate it from 1 to 10, or simply upvote/downvote the response) or inferred based
on user behavior (did the curse out the bot or ask it to modify the SQL after
initial generation). We'll ignore this feedback component for now, but in
practice this would be a crucial piece of the system.

> Additionally, we might bootstrap the system with a set of known English and SQL

Reviewer 2: You haven't mentioned what bootstrap is up to here, I think a like to what bootstrapping is would be helpful here.

```{python}
sql = text_to_sql(
    example_1.english,
    table_name=penguins.get_name,
    table_schema=penguins.schema,
    table_preview=str(penguins),
)
ibis.to_sql(penguins.sql(sql))
```

```{python}
penguins.sql(sql).limit(3)
```

## Modified example

```{python}
sql = text_to_sql(
    english,
    table_name=penguins.get_name,
    table_schema=penguins.schema,
    table_preview=str(penguins),
)
ibis.to_sql(penguins.sql(sql))
```

```{python}
penguins.sql(sql).limit(3)
```

:::

Despite low similarity, these English queries are effectively the same. Using
our embedding similarity RAG, we might miss this example as it could be under
the similarity threshold we set. Yet, a LLM can tell that these are semantically
equivalent, as shown by the SQL it generates for each.

## Discussion

I am by no means an expert on any of this. I studied neural networks to an
extent in college and I've kept up with all the hype around LLMs, but I'm not a
researcher nor a practitioner building products.

That said, I have noticed a pattern of tech-debt-inducing code for new machine
learning technologies when implemented in practice. I hope this blog has
provoked some thoughts around:

> - using bespoke databases when you may be using one that works just fine for the

Reviewer 2: ```suggestion
- Using bespoke databases when you may be using one that works just fine for the
```

 job to be done

> - using embeddings and similarity search for semantic ranking of text

Reviewer 2: ```suggestion
- Using embeddings and similarity search for semantic ranking of text
```

> - alternative approaches for getting the right context into your language models

Reviewer 2: ```suggestion
- Alternative approaches for getting the right context into your language models
```

Also, keep in mind that we're making external API calls to OpenAI for the

> embeddings. This may be costly (though embeddings calls are relatively cheap)

Reviewer 2: Might be worth providing an estimate on how much it cost you to run this, so people have a reference.
